server/src/test/java/io/crate/expression/scalar/bitwise/BitwiseFunctionsTest.java

```diff
@@ -0,0 +1,134 @@
+
+        // For computing expected values.
+        Map<String, TriFunction<DataType<Number>, Number, Number, Number>> referenceOperators = new HashMap<>();
+        referenceOperators.put("&", (t, a, b) -> t.implicitCast(a.longValue() & b.longValue()));
+        referenceOperators.put("|", (t, a, b) -> t.implicitCast(a.longValue() | b.longValue()));
+        referenceOperators.put("^", (t, a, b) -> t.implicitCast(a.longValue() ^ b.longValue()));
+        referenceOperators.put("#", (t, a, b) -> t.implicitCast(a.longValue() ^ b.longValue())); // # is overload of xor
+
+        for (DataType type : types) {
+            Supplier<Number> generator = DataTypeTesting.getDataGenerator(type); // Returns only not-null values.
+            for (Map.Entry<String, TriFunction<DataType<Number>, Number, Number, Number>> entry: referenceOperators.entrySet()) {
+                var a = generator.get();
+                var b = generator.get();
+                String expression = String.format(
+                    Locale.ENGLISH,
+                    "%d::%s %s %d::%s",
+                    a, type.getName(), entry.getKey(), b, type.getName()
+                );
+                String errorMessage = String.format(Locale.ENGLISH, "Error on computing expression %s", expression);
+                assertEvaluate(expression, entry.getValue().apply(type, a, b), errorMessage);
+            }
+        }
+    }
+
+    @Test
+    public void test_all_bitwise_operators_on_bit_string() {
+        // For computing expected values
+        Map<String, BinaryOperator<BitString>> referenceOperators = new HashMap<>();
+
+        referenceOperators.put("&", (bs1, bs2) -> {
+            bs1.bitSet().and(bs2.bitSet());
+            return bs1;
+        });
+        referenceOperators.put("|", (bs1, bs2) -> {
+            bs1.bitSet().or(bs2.bitSet());
+            return bs1;
+        });
+        referenceOperators.put("^", (bs1, bs2) -> {
+            bs1.bitSet().xor(bs2.bitSet());
+            return bs1;
+        });
+        referenceOperators.put("#", (bs1, bs2) -> {
+            bs1.bitSet().xor(bs2.bitSet());
+            return bs1;
+        });
+
+        var bitType = new BitStringType(randomIntBetween(1,100));
+        Supplier<BitString> generator = DataTypeTesting.getDataGenerator(bitType);
+        for (Map.Entry<String, BinaryOperator<BitString>> entry: referenceOperators.entrySet()) {
+            var a = generator.get();
+            var b = generator.get();
+            String expression = String.format(
+                Locale.ENGLISH,
+                "%s %s %s",
+                a.asPrefixedBitString(), entry.getKey(), b.asPrefixedBitString()
+            );
+            String errorMessage = String.format(Locale.ENGLISH, "Error on computing expression %s", expression);
+            assertEvaluate(expression, entry.getValue().apply(a, b), errorMessage);
+        }
+    }
+
+    @Test
+    public void test_parentheses_ensure_order() {
```

Comment:
not sure it's needed, more like testing parser

Reply:
Yes, please remove this here, it's up to the parser. If not tested already, please add it as a parser test instead.